[BUGFIX] Ne pas utiliser les traitements t1,t2,t3 pour les menu déroulants (PIX-2926).

MelanieMEB opened the pull request:
## :unicorn: Problème
Dans des épreuves QROC*, les traitements t1,t2,t3 peuvent changer la réponse de l'utilisateur et valider du texte avec quelques erreurs.
Par contre, dans le cas des menu déroulants, on a des choix données. Si les choix sont "Choix1", "Choix2", avec t3 activé (possibilité de faire une erreur), alors ces deux choix ont la même valeur.

## :robot: Solution
- Ne pas passer sur les traitements t1, t2 ou t3 si on sait que le bloc étudié est un menu déroulant
- Dans la classe solution, ajout d'un attribut `qrocBlocksType` avec le type de chaque bloc pour savoir si on utilise les traitements t*.

## :rainbow: Remarques

## :100: Pour tester
3 types d'épreuves : 
- QROCM Ind : rec2K0jwrN4moB80l
- QROCM Dep : recl3nvm0obJiDhGU
- QROC : reckyXWdHUHJLpury

MelanieMEB committed:
[API] Add a attributes qrocBlocksType to Solution to get the type of each qroc block

MelanieMEB committed:
[API] Refacto : send an object to QOCMInd validator with the full solution object

MelanieMEB committed:
[API] Use qrocBlocksType to ignore t1,t2,T3 when threated qrocm-ind challenge

pix-service commented:
I'm deploying this PR to these urls:

- App (.fr): https://app-pr3300.review.pix.fr
- App (.org): https://app-pr3300.review.pix.org
- Orga: https://orga-pr3300.review.pix.fr
- Certif: https://certif-pr3300.review.pix.fr
- Admin: https://admin-pr3300.review.pix.fr
- API: https://api-pr3300.review.pix.fr/api/

 Please check it out!

MelanieMEB committed:
[API] Send an object with all the solution object to QROCMDep validator

MelanieMEB committed:
[API] Use qrocBlocksType to ignore t1,t2,T3 when threated qrocm-dep

- Need to modify solutionServiceUtils to return value without modification for select qroc
- Add a missing possibility in the method _applyTreatmentsToSolutions

MelanieMEB committed:
[API] Send an object with all the solution object to QROC validator

MelanieMEB committed:
[API] Use qrocBlocksType to ignore t1,t2,T3 when threated qroc

VincentHardouin commented on api/tests/unit/domain/services/solution-service-utils_test.js:
```diff
@@ -1,36 +1,46 @@
     });
+
+    context('when we do not want apply treatments', () => {
+      it('should return the same answer withou treatment', () => {
```
```suggestion
      it('should return the same answer without treatment', () => {
```

francois2metz commented:
Tu aurais une épreuve avec menu déroulant et un choix en t{1,2,3} ?

sbedeau commented on api/lib/infrastructure/adapters/solution-adapter.js:
```diff
@@ -8,11 +8,28 @@ function statusToBoolean(value) {
+function _extractTypeOfQroc(datasourceObject) {
+  if (datasourceObject.type === 'QCU' || datasourceObject.type === 'QCM') {
+    return {};
+  } else {
```
**suggestion:** Enlever le `else` qui est redondant de par le `return` en ligne 13

sbedeau commented on api/lib/infrastructure/adapters/solution-adapter.js:
```diff
@@ -8,11 +8,28 @@ function statusToBoolean(value) {
+
+    Array.from(qrocBlocks, (qrocBlock) => {
+      const qrocBlockKey = qrocBlock[0].replace('${', '').match(/^(.+?)(#|§|}| options)+/i)[1];
+      const qrocBlockKType = qrocBlock[1].includes('options=') ? 'select' : 'text';
+      qrocBlocksTypes[qrocBlockKey] = qrocBlockKType;
```
**chipotage:** Il y a une typo sur `qrocBlockKType`

sbedeau commented on api/lib/infrastructure/adapters/solution-adapter.js:
```diff
@@ -8,11 +8,28 @@ function statusToBoolean(value) {
+    const qrocBlocks = datasourceObject.proposals ? datasourceObject.proposals.matchAll(/\$\{(.*?)\}/g) : [];
+
+    Array.from(qrocBlocks, (qrocBlock) => {
+      const qrocBlockKey = qrocBlock[0].replace('${', '').match(/^(.+?)(#|§|}| options)+/i)[1];
```
**suggestion:** Expliciter les regex par des appels de méthodes

sbedeau commented on api/tests/unit/domain/services/solution-service-utils_test.js:
```diff
@@ -1,36 +1,46 @@
 const service = require('../../../../lib/domain/services/solution-service-utils');
 
-describe('Unit | Domain | Services | solution-service-utils', function() {
+describe('Unit | Domain | Services | solution-service-utils', () => {
```
**chipotage:** La PR sur #3310 force l'utilisation de `function()` au lieu de `() =>`

sbedeau reviewed (approved):
Vraiment pas simple cette portion du code, bien joué ! 
J'ai aussi besoin d'une épreuve, comme François

MelanieMEB commented:
J'ai pris les commentaires en compte, mais j'ai relevé un soucis sur les QROCM-dep : 
- Vu que tous les inputs peuvent donner des valeurs pour la solution, les traitements t1/t2 sont fait sur les solutions (ex : la bonne réponse est Choix1 et on la change donc en choix1), mais pas sur les entrées des Select. Donc un cas un peu spécifique ici s'il y a une majuscule dans les select par exemple (car le traitement aura retiré la majuscule coté solution, mais pas coté réponse, et on a besoin du traitement pour le champ "input" qui pourrait aussi y avoir). Cas à la marge quand même.

MelanieMEB committed:
[API] Refacto step 1 : simply two same method and use the same name

MelanieMEB committed:
[API] Refacto step 2 : create a method to apply treatments with deactivations

MelanieMEB committed:
[API] Refacto step 3: use one method to apply treatments

VincentHardouin commented on api/lib/domain/services/solution-service-qroc.js:
```diff
@@ -48,42 +50,20 @@ function _getAnswerStatusFromNumberMatching(answer, solutions) {
-    }
-    else if (deactivationsService.hasT1T2T3(deactivations)) {
+    const shouldIgnoreAllTreatments = ignoreTreatments.length === 3;
+    if (shouldIgnoreAllTreatments) {
```
`applyTreatments` sait déjà s'il doit désactiver où non des traitements. Je me demande si c'est pas plus claire que les `ignoreTreatments` deviennent des `deactivations` au lieu de rajouter une couche à cet endroit. 🤔

MelanieMEB committed:
[API] Refacto step 4: only use applyTreatments variable to the applyTreatments method

MelanieMEB committed:
[API] Remove ignoring treatment for qrocm-dep

VincentHardouin commented on api/lib/infrastructure/adapters/solution-adapter.js:
```diff
@@ -8,11 +8,36 @@ function statusToBoolean(value) {
+  const qrocBlocksTypes = {};
+  const qrocBlocks = _getAllBlocks(datasourceObject.proposals);
+
+  Array.from(qrocBlocks, (qrocBlock) => {
```
`qrocBlocks` est toujours un tableau comme `_getAllBlocks()` retourne toujours un tableau.
Du coup, je me demande si c'est nécessaire de créer une copie de `qroBlocks` ici ? J'ai l'impression qu'on souhaite principalement le parcourir.

MelanieMEB replied:
J'ai modifié _getAllBlocks qui ne renvoyait pas toujours un tableau pour faire un simple forEach ici.

ghost commented:
C'est assez complexe tout ça mais je n'ai rien vu de choquant en lecture simple.
À voir à l'usage si on a des problèmes.

MelanieMEB committed:
[API] Lint fix
